tests/devices/default_qubit/test_default_qubit.py

```diff
@@ -1637,6 +1638,268 @@ def test_projector_dynamic_type(max_workers, n_wires):
+    )
+    @pytest.mark.parametrize("param", np.linspace(np.pi / 4, 3 * np.pi / 4, 3))
+    @pytest.mark.parametrize("shots", [10, (10, 10)])
+    def test_postselection_valid_finite_shots_varied_shape(
+        self, mp, param, expected_shape, shots, interface, use_jit
+    ):
+        """Test that qml.sample and qml.classical_shadow work correctly.
+        Separate test because their shape is non-deterministic."""
+
+        if use_jit:
+            pytest.skip("Cannot JIT while mocking function.")
+
+        dev = qml.device("default.qubit", seed=42)
+        param = qml.math.asarray(param, like=interface)
+
+        with mock.patch("numpy.random.binomial", lambda *args, **kwargs: 5):
+
+            @qml.qnode(dev, interface=interface)
+            def circ_postselect(theta):
+                qml.RX(theta, 0)
+                qml.CNOT([0, 1])
+                qml.measure(0, postselect=1)
+                return qml.apply(mp)
+
+            if use_jit:
+                import jax
+
+                circ_postselect = jax.jit(circ_postselect, static_argnames=["shots"])
+
+            res = circ_postselect(param, shots=shots)
+
+        if not isinstance(shots, tuple):
+            assert qml.math.get_interface(res) == interface if interface != "autograd" else "numpy"
+            assert qml.math.shape(res) == expected_shape
+
+        else:
+            assert isinstance(res, tuple)
+            for r in res:
+                assert (
+                    qml.math.get_interface(r) == interface if interface != "autograd" else "numpy"
+                )
+                assert qml.math.shape(r) == expected_shape
+
+    @pytest.mark.parametrize(
+        "mp, autograd_interface, is_nan",
+        [
+            (qml.expval(qml.PauliZ(0)), "autograd", True),
+            (qml.var(qml.PauliZ(0)), "autograd", True),
+            (qml.probs(wires=[0, 1]), "autograd", True),
+            (qml.state(), "autograd", True),
+            (qml.density_matrix(wires=0), "autograd", True),
+            (qml.purity(0), "numpy", True),
+            (qml.vn_entropy(0), "numpy", False),
+            (qml.mutual_info(0, 1), "numpy", False),
+        ],
+    )
+    def test_postselection_invalid_analytic(
+        self, mp, autograd_interface, is_nan, interface, use_jit
+    ):
+        """Test that the results of a qnode are nan values of the correct shape if the state
+        that we are postselecting has a zero probability of occurring."""
+
+        if (isinstance(mp, qml.measurements.MutualInfoMP) and interface != "jax") or (
+            isinstance(mp, qml.measurements.VnEntropyMP) and interface == "tensorflow"
+        ):
+            pytest.skip("Unsupported measurements and interfaces.")
+
+        if use_jit and interface != "jax":
+            pytest.skip("Can't jit with non-jax interfaces.")
+
+        # Wires are specified so that the shape for measurements can be determined correctly
```

Comment:
i'm still worried about this. if we advertise that you don't need to specify device wires, this should include postselection too. doesn't the device preprocessing step put wires on measurements?

Reply:
This is just for the test because I'm comparing the shape of the output to `mp.shape`. Using `StateMP.shape` with `DefaultQubit` without wires fails because it uses device wires to figure out what the shape of the output should be. And device pre-processing only applies to qnodes, not to tapes, but that's not really relevant here.

Reply:
After my follow up PR `qml.defer_measurements` will not support `qml.state` at all.